### DISCUSSION TITLE
Plugin ExecFilterCommand - not working?

### DISCUSSION THREAD

**Discussion by clach04** [WinMerge/winmerge#2414] [Q&A] [answered]
I know WinMerge has binary diff support BUT I'm looking to diff some encrypted files. As an initial test, I thought I'd try and use xxd from vim (or any other hex dump tool)  to do a test hex diff without generating temporary disk files.

    ExecFilterCommand xxd %1

Auto plugin config doesn't work. The plugin is never run, even when setting a wild card to match the file extension

Manually applying the plugin needs to be run for each side/pane. By default, it seems to retain the binary view rather than operating as a filter as it autoselects the diff not the entire file :-( If you manually select all the content then manually apply (PLugins, Scripts, Apply Filter Command the output will be replaced BUT then there is no diff :-(.

Also when manually running, the "Plugin Settings" button generates an error pop-up, "Unknown name". This part feels like a bug.

I'm not sure if the other behaviors are a bug or user problem. The documentation at https://manual.winmerge.org/en/Plugins.html (and the help file) don't have any examples.

My test is to generate 2 small binary files with the extension `.biny` so as to not trigger the hex diff viewer that's built in.

Generated via:

    #!/usr/bin/env python3
    """Generate tiny binary file.

    Convert integers to single-byte value.

    Use Hex dump tool, like:
      * hex_head.py
      * xxd
      * od
      * hex_dump, etc.

    Compare tests

        C:\code\>diff tinydemo.bin tinydemo_modified.bin
        Binary files tinydemo.bin and tinydemo_modified.bin differ

        C:\code\>fc tinydemo.bin tinydemo_modified.bin
        Comparing files tinydemo.bin and TINYDEMO_MODIFIED.BIN
        00000041: 41 44
        00000042: 42 45
        00000043: 43 41
        00000045: 45 42
        00000046: 46 45
        00000047: 47 45
        00000048: 48 46

    Winmerge

    default, built in:

        WinMergeU tinydemo.bin tinydemo_modified.bin

    rename to skip built in hex diff

        ren tinydemo.bin tinydemo.bin y
        ren tinydemo_modified.bin tinydemo_modified.biny
        WinMergeU tinydemo.biny tinydemo_modified.biny

    ExecFilterCommand 

        ExecFilterCommand xxd %1

    """

    result = []
    for x in range(0xff + 1):
        print(x)
        result.append(x.to_bytes())

    result_bytes = b''.join(result)
    f = open('tinydemo.bin', 'wb')
    f.write(result_bytes)
    f.close()

    result_bytes = result_bytes.replace(b'ABCDEFGH', b'DEADBEEF')
    f = open('tinydemo_modified.bin', 'wb')
    f.write(result_bytes)
    f.close()

**Comment by sdottaka** [comment, marked as answer]
First, WinMerge does not apply plugins automatically by default, so you need to select the Plugins → Automatic Unpacking menu item as shown below to apply the plugin automatically.

Recent versions of WinMerge allow you to create plugin aliases and simple plugins, so try creating one as follows.

1. Compare the *.biny files.

![image](https://github.com/user-attachments/assets/f6ef23da-bb8e-4958-907c-988de79b3b6e)

2. Select Plugins → Edit with Unpacker menu item.

![image](https://github.com/user-attachments/assets/39a7b679-3a85-404e-a133-3cb59bf8e2d8)

3. Enter the following and press the Alias... button.

![image](https://github.com/user-attachments/assets/8e17792a-3787-4967-aa7c-9c4dfd52d461)

4. Enter the following.

![image](https://github.com/user-attachments/assets/6896846e-7848-4c21-9343-455b237c3f72)

5. Compare the *.biny files again and the plugin should be applied.

![image](https://github.com/user-attachments/assets/7c43049b-0f43-4dfb-ac01-f3c5fcddb20f)

Any aliases you create can be deleted in the Plugins window.

![image](https://github.com/user-attachments/assets/60ad37a7-82a3-4ca7-b3f6-9ec309a0ac9c)
